source/app/models/url.rb

```diff
@@ -0,0 +1,14 @@
+class Url < ActiveRecord::Base
+  
+  before_save :shorten_url
+  validates_uniqueness_of :unique_key 
+
+  private
+    #generate 6 digit unique identifier
+    def shorten_url(record)
+      if self.unique_key.nil?
+        self.unique_key = 6.times.map { [*'0'..'9', *'A'..'Z'].sample }.join
```

Comment:
This array of characters you're calling `sample` on can be extracted to a constant. That way ruby doesn't have to reconstruct it 6 times for every call to this method.